docs/developer/development-lifecycle.md

```diff
@@ -0,0 +1,33 @@
+   1. Do we need to solve this problem and if so, why? If a feature advanced all the way to a GitHub issue, the answer most of the time will be yes. But don't fear to question the requirements. It's important to understand not only what we are building but why, or we might end up implementing the wrong thing. If you don't think this problem needs to be solved, it's already been solved or there a better way, this is the moment to voice it out.
+   2. What are we building? Sometimes the issue will be prescriptive about the solution to the problem, sometimes it will just give you functional requirements and other times you will only have a description of the problem. Make sure you understand all of this and collect all the requirements before moving forward with the design. This will sometimes involve talking to other engineers, product or even users. If you disagree with any given requirement, challenge them.
+1. If the task needs a design, work on it before starting the implementation.
+   * Not all features need a design doc, but most will need a design. Study problem space and the existing code and come up with a solution before implementing it.
+   * When possible, favor getting eyes on your ideas/design/doc as soon as possible. Early feedback avoids rework and will set you up for a speedy review once you write the code. Not everything needs a formal design review on a PR, just ask folks for a chat to bounce ideas, specially folks with special experience or interest in the are you are working on. This is specially important when introducing new ideas or patterns.
+   * Paper sustains everything so if you think you need to prove an idea, feel free to play with a POC. The goal of designing the solution before implementing is to avoid rework and get early feedback, code is as valid as any other tool to come up with that design. That said, most features won't need a POC.
+   * Timebox this, specially for big features. If it's taking too long (relative to the size of the task) you might be trying to go too deep. In that case, ask someone to pair on the design, it's usually easier to stay focused on what is needed when you can keep each other accountable.
+   * Favor speed over perfection. Look for two way doors and if a decision can be postponed without adding risk, leave it for later.
+   * Make sure you include testing in your design. When a feature can benefit from E2E tests, it can be useful to enumerate the scenarios that need to be tested.
+1. If needed, split the task into manageable chunks. Feel free to create new GitHub issues. You can track them in the original one with a checklist or you can even create a GitHub project (specially useful for big features, even more so when they span across multiple releases). This helps to keep track of the progress and facilitates adding more people to the feature if needed.
```

Comment:
What's our preferred method for delivering incrementally? Might be worth a call out to prefer smaller scoped PRs delivered over time with feature flags versus giant ones that implement everything in one go and are hard to rollback.